### PR TITLE
login: don't remove all devices from PID1 when only one was removed

### PR DESCRIPTION
**Pull request by sourcejedi**
Fix for #8344.

### CODE REVIEW

File: src/login/logind-session-device.c [hunk 1]
```diff
@@ -417,14 +417,20 @@ void session_device_free(SessionDevice *sd) {
...
+                _cleanup_free_ char *fdname = NULL;
                 const char *m;
+                int r;
 
                 /* Make sure to remove the pushed fd. */
 
-                m = strjoina("FDSTOREREMOVE=1\n"
-                             "FDNAME=session-", sd->session->id);
+                r = asprintf(&fdname, "device-%u-%u-session-%s",
+                                      major(sd->dev), minor(sd->dev), sd->session->id);
```
Comment (keszybz) [inline src/login/logind-session-device.c]: I'd prefer this to be "session-%s-device-%u-%u".

File: src/login/logind-session-device.c [hunk 2]
```diff
@@ -417,14 +417,20 @@ void session_device_free(SessionDevice *sd) {
...
+                if (r < 0) {
```
Comment (keszybz) [inline src/login/logind-session-device.c]: Hmm, `r < 0`? I think you mean `r >= 0`.

I don't think it makes much sense to do asprintf and then strjoina. Just asprintf the whole thing.

File: src/login/logind-session-device.c [hunk 3]
```diff
@@ -525,8 +532,13 @@ int session_device_save(SessionDevice *sd) {
...
                 return 0;
 
+        r = asprintf(&fdname, "device-%u-%u-session-%s",
+                              major(sd->dev), minor(sd->dev), sd->session->id);
+        if (r < 0)
+                return r;
+
         m = strjoina("FDSTORE=1\n"
-                     "FDNAME=session-", sd->session->id);
+                     "FDNAME=", fdname, "\n");
```
Comment (keszybz) [inline src/login/logind-session-device.c]: Here too.

File: src/login/logind.c [hunk 4]
```diff
@@ -411,6 +412,43 @@ static int manager_enumerate_users(Manager *m) {
         return r;
 }
 
+static int parse_fdname(const char *fdname, char **id, dev_t *dev)
+{
```
Comment (keszybz) [inline src/login/logind.c]: Please move { to the previous line.

File: src/login/logind.c [hunk 5]
```diff
@@ -411,6 +412,43 @@ static int manager_enumerate_users(Manager *m) {
...
+
+        r = safe_atou(parts[1], &major);
+        if (r < 0)
+                return r;
+        r = safe_atou(parts[2], &minor);
+        if (r < 0)
+                return r;
+
+        if (!streq(parts[3], "session"))
+                return -EINVAL;
```
Comment (keszybz) [inline src/login/logind.c]: This looks like a sscanf would be easier than doing the parsing by hand.

File: src/login/logind.c [hunk 6]
```diff
@@ -427,13 +465,16 @@ static int manager_attach_fds(Manager *m) {
                 struct stat st;
                 SessionDevice *sd;
                 Session *s;
-                char *id;
+                dev_t dev;
+                _cleanup_free_ char *id;
```
Comment (keszybz) [inline src/login/logind.c]: This needs to be initalized, it's used in the failure path below.

File: src/login/logind-session-device.c (hunk 2 above)
Comment (keszybz) [reply src/login/logind-session-device.c]: Or do
```C
if () {
    char buf[strlen("FDSTOREREMOVE=1\n" "FDNAME=session--device--") + strlen(sd->session->id) + 2*DECIMAL_STR_MAX(unsigned) + 1];
    ...
    xsprintf(buf, ...);
```
This will avoid the ugly error handling of failing asprintf.

File: src/login/logind.c (hunk 5 above)
Comment (sourcejedi) [reply src/login/logind.c]: Thanks for your review!  I'm thinking of sticking with the "by hand" approach here.

%s is greedy and consumes all non-whitespace, so it would not work after changing to `session-%s-device-%d-%d`.  *scanf() also makes people nervous due to the insecure design of %s.  With sscanf() you can work around it but it's still hacky.

I since noticed

> The session id was put at the end, just in case it contained a separator

was not necessarily needed, so the fun detail with strjoin() below can go away.

**Comment by sourcejedi** [comment]
I've reworked the main commit.  It should meet the above comments, with the exception that I have not switched to sscanf().

I also fixed several other bugs in my commit.  Sloppy of me.

I was hoping for just a little bikeshedding; your comments were rather helpful.  Thanks again for reviewing :).

File: src/login/logind.c (hunk 4 above)
Comment (yuwata) [reply src/login/logind.c]: Not fixed yet.

**Comment by sourcejedi** [comment]
Right.  Here's that style fix applied to the current version.  (As opposed to the version I threw away :).

File: src/login/logind-session-device.c [hunk 7]
```diff
@@ -413,16 +413,36 @@ int session_device_new(Session *s, dev_t dev, bool open_device, SessionDevice **
         return r;
 }
 
+static int snprintf_fdname(char *buf, int size, SessionDevice *sd) {
+        const char *p;
+        int r;
+
+        /* Session ID does not contain separators. */
+        for (p = sd->session->id; *p; p++)
+                assert(!IN_SET(*p, '-', '\n'));
```
Comment (keszybz) [inline src/login/logind-session-device.c]: Loops like this are surprisingly slow. See https://github.com/systemd/casync/commit/b1cc17ba25f66a489fe31bfd72c8dac6c7649192 for an extreme case in casync.

This would be better written as
```C
        assert((sd->session->id + strcspn(sd->session->id, "-\n")) == '\0');
```

File: src/login/logind-session-device.c [hunk 8]
```diff
@@ -413,16 +413,36 @@ int session_device_new(Session *s, dev_t dev, bool open_device, SessionDevice **
...
+                char *fdname;
                 const char *m;
 
                 /* Make sure to remove the pushed fd. */
 
-                m = strjoina("FDSTOREREMOVE=1\n"
-                             "FDNAME=session-", sd->session->id);
+                size = snprintf_fdname(NULL, 0, sd);
+                fdname = alloca(size);
+                (void) snprintf_fdname(fdname, size, sd);
```
Comment (keszybz) [inline src/login/logind-session-device.c]: Sorry, but this is just overly verbose and slow. There really is no point in doing dynamic allocation and then copying that to a bigger stack-allocated buffer. Please try my suggestion in https://github.com/systemd/systemd/pull/8377#discussion_r172890823.

**Comment by sourcejedi** [comment]
Updated.  Optimization as suggested.  Minus 11 lines, good point.

**Comment by keszybz** [comment]
```
../src/login/logind-session-device.c: In function ‘session_device_save’:
../src/login/logind-session-device.c:538:9: error: ‘session_id’ undeclared (first use in this function)
         session_id = sd->session->id;
         ^
```

File: src/login/logind-session-device.c [hunk 9]
```diff
@@ -416,15 +416,21 @@ int session_device_new(Session *s, dev_t dev, bool open_device, SessionDevice **
...
         if (sd->pushed_fd) {
-                const char *m;
-
-                /* Make sure to remove the pushed fd. */
-
-                m = strjoina("FDSTOREREMOVE=1\n"
-                             "FDNAME=session-", sd->session->id);
-
-                (void) sd_notify(false, m);
+                _cleanup_free_ char *m;
```
Comment (keszybz) [inline src/login/logind-session-device.c]: `m` needs to be initailzied.

File: src/login/logind-session-device.c [hunk 10]
```diff
@@ -510,7 +516,8 @@ unsigned int session_device_try_pause_all(Session *s) {
 }
 
 int session_device_save(SessionDevice *sd) {
-        const char *m;
+        _cleanup_free_ char *m;
```
Comment (keszybz) [inline src/login/logind-session-device.c]: Here too.

**Comment by sourcejedi** [comment]
You're a saint.  Fixed.

**Comment by keszybz** [comment]
> You're a saint. Fixed.

Finally! Proper appreciation for my grumpy comments ;)

LGTM. Thanks for working on this.

**Comment by keszybz** [comment]
```
autopkgtest login: 
---------------------------------------------------
sudo: /tmp/autopkgtest-run-wrapper: command not found
Exit request sent.
<VirtSubproc>: failure: timed out waiting for testbed to reboot
autopkgtest [13:39:15]: ERROR: testbed failure: cannot send to testbed: [Errno 32] Broken pipe
```

/cc @martinpitt

**Comment by fbuihuu** [comment]
```
login: don't remove all devices from PID1 when only one was removed
[...]
Upgrading logind on a running system with a GUI might start being possible
after this commit (and after also fixing the display server of your
choice).
```
@sourcejedi : hmm since commit aed24c4cd7641da6f530853d10637568c13c8f35, restarting logind with a GUI is supposed to work, at least that's what my testing showed. Therefore I'm not sure what you meant here...

I'm asking because I'd like to know if some commits really need to be backported on the version used by Tumbleweed.

**Comment by sourcejedi** [comment]
Hi, nice to talk to you!

1. In stock Fedora 27 (systemd v234), restarting logind asserted. #8035.  If you just null out the assert, there was a second one, hit when you switch _to_ a VT with an active graphics server. #8343.

   The root cause for at least the first one, was that we set `FDSTORE=session{$XDG_SESSION_ID}`, but what we expect on the restart is `FDSTORE=session-{$XDG_SESSION_ID}`. Fixed in PR #8345 / commit https://github.com/systemd/systemd/pull/8345/commits/b5cdfa40ca6aae5e98a69da7a8b350215f807649

   The other effect of this root cause, is that DRM fds are leaked in PID1 and never closed.  This effectively causes [real big memory leaks](https://unix.stackexchange.com/questions/431982/what-could-be-using-6gb-of-my-swap) when a graphics server shuts down uncleanly and hasn't de-allocated resources :).

2. In this PR #8377, we observe that removing one session device caused all devices for that session to be removed from PID1.  Therefore restarting logind will cause it to lose device fds, which basically breaks switching between different unprivileged graphics servers.

   I think I hit device removal constantly in a VM - maybe as often as when the mouse cursor leaves and re-enters the VM window?  #8343 above might also have been related to this.

3. Xorg is still so far choosing to die, when logind is stopped or restarted.  So anyone using Xorg will not be restarting logind. `gnome-shell` does not choose to die, but unfortunately it does not manage to re-open input devices correctly.  gnome [#1540321](https://bugzilla.redhat.com/show_bug.cgi?id=1540321).  Therefore anyone using gnome-shell will almost certainly not be restarting logind.  Therefore I didn't think about how to support restarting into a logind with the above fixes, because people will have already not been restarting logind when they upgrade it.  And hopefully _since_ the above fixes, everything now works,  and people can start thinking if they want to support restarting logind, in graphics servers and in systemd upgrade scenarios.

**Comment by sourcejedi** [comment]
Don't tell me, I can guess the most annoying possibility :).  KDE's implementation of Wayland managed to survive logind restarts without losing input or drm fds?  I can imagine you might not notice the rest if you weren't doing *any* VT switching.  (I get into the habit of VT switching because I do unpriviliged web browsing in a different user account).

I guess you'd want to drop the parsing of major/minor in the FDNAME (the new structure still needs to stay on the FDSTORE=1 side), accept both "FDSTORE=session-{$XDG_SESSION_ID}" and "FDSTORE=session-{$XDG_SESSION_ID}-*", and read the device major/minor from `fstat()` instead.

I'm working on logind again right now, I wouldn't really mind sending you a compile-tested patch for you to test KDE with if you like.

**Comment by sourcejedi** [comment]
(In the theoretical future it would be nicer if we _didn't_ need to rely on the PID1 fdstore, and we just had something like the "stealmaster" mentioned for DRM, and a [revoke()](https://lwn.net/Articles/546537/) or equivalent that at least worked on input devices).

**Comment by fbuihuu** [comment]
@sourcejedi thanks for your prompt answer !

If I'm not mistaken, 1. was due to logind being killed due to a bug in KVM. Even if it should be fixed that shouldn't be a problem in the common case, right ?

Regarding the naming error (with FDSTORE=session) has been introduced later by commit 4050e4797603d3644707d58edfd9742b5311c7cf which is part of v236.

Regarding #8343 it's been detected on v237+.

So even if v234 has some issues in some corner cases, restarting should already work if I'm not mistaken.

Again I'm asking because I'd like to know if I need to consider backporting some major fixes on top of v234.

BTW that sad that all this happened without actually being notified.

**Comment by sourcejedi** [comment]
The KVM bug is why logind was restarted.  However restarting logind normally also caused the first assert.

> Regarding the naming error (with FDSTORE=session) has been introduced later by commit 4050e47 which is part of v236.

I see, thanks!

> Regarding #8343 it's been detected on v237+.

Right... I think #8344 is caused by the issue this PR #8377 fixes, so it would also affect v234, but it would be most visible inside a VM.  Or if you use a laptop dock or something.

> So even if v234 has some issues in some corner cases, restarting should already work if I'm not mistaken.

Fair, except I don't have or know for sure any _client_ of logind devices (graphics server) which survives logind restart intact, for me to test it.  I.e. point 3).  I'm only guessing that you had a KDE that continued working.  Can you confirm my guess?  Do you want to test a patch as suggested?

**Comment by sourcejedi** [comment]
> BTW that sad that all this happened without actually being notified.

I was also sad about the response to my first two bug reports.  I did come across your commit later on, but it's not always clear whether people are still interested in bug reports.  This may be the first time my PRs would have benefited from CC'ing an original author - I will try to remember that in mind in future.

(I.e. when I patch a defect in code introduced by poettering or keszybz, likely either they're the reviewer for my PR, or they reviewed the original commit, and I feel like I'm yelling if I make a point of always attributing blame).

**Comment by fbuihuu** [comment]
> Fair, except I don't have or know for sure any client of logind devices (graphics server) which survives logind restart intact, for me to test it. I.e. point 3). I'm only guessing that you had a KDE that continued working. Can you confirm my guess? Do you want to test a patch as suggested?

For testing purposes we reverted the commit that made X exit when it detects that logind was restarted otherwise we couldn't have made any useful testing ;) see https://github.com/systemd/systemd/pull/5600#issuecomment-310870680

**Comment by fbuihuu** [comment]
> I was also sad about the response to my first two bug reports. I did come across your commit later on, but it's not always clear whether people are still interested in bug reports. This may be the first time my PRs would have benefited from CC'ing an original author - I will try to remember that in mind in future.

I wasn't blaming you in any cases. It's just a pity that there's no way to be notified when a part you worked on is fixed or improved somehow.

**Comment by sourcejedi** [comment]
Ah, thanks for clarification!  I think the traditional notification mechanism is when the maintainer copies you in because they think you might know the answer :).

It sounds like you don't need to worry too much about logind restarts, unless you have _packaged_ a version of Xorg that tolerates the restart of logind?

Personally I would still want to make sure I avoided the bug which leaks DRM fds, causing large memory leaks in case of unclean X shutdown.  Because it makes annoying crashes significantly more annoying.   So if you end up packaging v236 or v237, you might want to look at PR #8345 / commit b5cdfa4.

I think the other issues I worked on all involved a logind restart.

**Comment by fbuihuu** [comment]
> It sounds like you don't need to worry too much about logind restarts, unless you have packaged a version of Xorg that tolerates the restart of logind?

Nope. But better to be prepared when it will be done.

Thanks.
